### PR TITLE
Fs heap strdup asprintf

### PR DESCRIPTION
**Pull request by crafcat7**
## Summary

**Why is this change necessary?**

After https://github.com/apache/nuttx/pull/13722, we found that some variables in vfs are stored in memory through strdup / asprintf. In order to keep the original intention of fs_heap as much as possible, we changed their usage to fs_heap_xxx to avoid the potential allocation and release in different heap regions.

**Changes**

We searched for all uses of strdup / asprintf in nuttx/fs and changed them to equivalent functions through fs_heap_malloc

## Impact

**Is a new feature added?:** NO
**Impact on build:** NO
**Impact on hardware:** NO, this change does not specifically target any particular hardware architectures or boards.
**Impact on documentation:** NO,This patch does not introduce any new features
**Impact on compatibility:** This implementation aims to be backward compatible. No existing functionality is expected to be broken.

## Testing

Build Host(s): Linux x86
Target(s): sim/nsh

It can pass in our FS related test cases

```
[==========] nuttx_fs_test_suites: Running 40 test(s).
[ RUN      ] test_nuttx_fs_creat01
[       OK ] test_nuttx_fs_creat01
[ RUN      ] test_nuttx_fs_dup01
[       OK ] test_nuttx_fs_dup01
[ RUN      ] test_nuttx_fs_fcntl01
[       OK ] test_nuttx_fs_fcntl01
[ RUN      ] test_nuttx_fs_fstat01
[       OK ] test_nuttx_fs_fstat01
[ RUN      ] test_nuttx_fs_fstatfs01
[       OK ] test_nuttx_fs_fstatfs01
[ RUN      ] test_nuttx_fs_fsync01
[       OK ] test_nuttx_fs_fsync01
[ RUN      ] test_nuttx_fs_fsync02
the bsize = 4096
[       OK ] test_nuttx_fs_fsync02
[ RUN      ] test_nuttx_fs_getfilep01
[       OK ] test_nuttx_fs_getfilep01
[ RUN      ] test_nuttx_fs_mkdir01
[       OK ] test_nuttx_fs_mkdir01
[ RUN      ] test_nuttx_fs_open01
[       OK ] test_nuttx_fs_open01
[ RUN      ] test_nuttx_fs_opendir01
[       OK ] test_nuttx_fs_opendir01
[ RUN      ] test_nuttx_fs_opendir02
[       OK ] test_nuttx_fs_opendir02
[ RUN      ] test_nuttx_fs_pread01
[       OK ] test_nuttx_fs_pread01
[ RUN      ] test_nuttx_fs_pwrite01
[       OK ] test_nuttx_fs_pwrite01
[ RUN      ] test_nuttx_fs_readdir01
[       OK ] test_nuttx_fs_readdir01
[ RUN      ] test_nuttx_fs_readlink01
[       OK ] test_nuttx_fs_readlink01
[ RUN      ] test_nuttx_fs_rename01
[       OK ] test_nuttx_fs_rename01
[ RUN      ] test_nuttx_fs_rename02
[       OK ] test_nuttx_fs_rename02
[ RUN      ] test_nuttx_fs_rewinddir01
[       OK ] test_nuttx_fs_rewinddir01
[ RUN      ] test_nuttx_fs_fmdir01
[       OK ] test_nuttx_fs_fmdir01
[ RUN      ] test_nuttx_fs_rmdir02
[       OK ] test_nuttx_fs_rmdir02
[ RUN      ] test_nuttx_fs_rmdir03
[       OK ] test_nuttx_fs_rmdir03
[ RUN      ] test_nuttx_fs_seek01
[       OK ] test_nuttx_fs_seek01
[ RUN      ] test_nuttx_fs_seek02
[       OK ] test_nuttx_fs_seek02
[ RUN      ] test_nuttx_fs_statfs01
[       OK ] test_nuttx_fs_statfs01
[ RUN      ] test_nuttx_fs_symlink01
[       OK ] test_nuttx_fs_symlink01
[ RUN      ] test_nuttx_fs_truncate01
[       OK ] test_nuttx_fs_truncate01
[ RUN      ] test_nuttx_fs_unlink01
[       OK ] test_nuttx_fs_unlink01
[ RUN      ] test_nuttx_fs_write01
[       OK ] test_nuttx_fs_write01
[ RUN      ] test_nuttx_fs_write02
[       OK ] test_nuttx_fs_write02
[ RUN      ] test_nuttx_fs_write03
[       OK ] test_nuttx_fs_write03
[ RUN      ] test_nuttx_fs_append01
[       OK ] test_nuttx_fs_append01
[ RUN      ] test_nuttx_fs_sendfile01
[       OK ] test_nuttx_fs_sendfile01
[ RUN      ] test_nuttx_fs_sendfile02
[       OK ] test_nuttx_fs_sendfile02
[ RUN      ] test_nuttx_fs_stream01
[       OK ] test_nuttx_fs_stream01
[ RUN      ] test_nuttx_fs_stream02
[       OK ] test_nuttx_fs_stream02
[ RUN      ] test_nuttx_fs_stream03
[       OK ] test_nuttx_fs_stream03
[ RUN      ] test_nuttx_fs_stream04
[       OK ] test_nuttx_fs_stream04
[ RUN      ] test_nuttx_fs_eventfd
[       OK ] test_nuttx_fs_eventfd
[ RUN      ] test_nuttx_fs_poll01
[       OK ] test_nuttx_fs_poll01
[==========] nuttx_fs_test_suites: 40 test(s) run.
[  PASSED  ] 40 test(s).
```

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



The PR summary and impact assessment are well-written and informative. However, the testing section could be improved:

* **More details on testing:**  Simply stating "It can pass in our FS related test cases" is not sufficient. Please provide more specific information about the tests conducted:
    * **Which test cases exactly?** List the names of the specific test cases you executed. 
    * **How were they run?** Did you use a specific test framework or script?
    * **Test environment:** Provide details like the board configuration used for sim target testing.
* **Logs are missing:** The template requests "Testing logs before change" and "Testing logs after change." Please include relevant snippets from the test output.  While a full log might be lengthy, focus on sections demonstrating the change's impact.
* **Consider edge cases:**  Did you test any scenarios where memory allocation from `fs_heap` might be stressed? This helps ensure the changes don't introduce memory leaks or fragmentation issues.

By addressing these points, you'll make it easier for reviewers to understand the extent and thoroughness of your testing, increasing confidence in your changes.

File: fs/driver/fs_blockproxy.c [hunk 1]
```diff
@@ -112,7 +115,16 @@ static FAR char *unique_chardev(void)
       if (ret < 0)
         {
           DEBUGASSERT(ret == -ENOENT);
-          return strdup(devbuf);
```
Comment (xiaoxiang781216) [inline fs/driver/fs_blockproxy.c:115]: let's add fs_heap_strdup?

File: fs/inode/fs_inodesearch.c [hunk 2]
```diff
@@ -347,21 +348,21 @@ static int _inode_search(FAR struct inode_search_s *desc)
                               if (*desc->relpath != '\0')
                                 {
                                   FAR char *buffer = NULL;
+                                  size_t bufsize = strlen(desc->relpath) +
+                                  strlen(name) + 2;
 
-                                  ret = asprintf(&buffer,
```
Comment (xiaoxiang781216) [inline fs/inode/fs_inodesearch.c:351]: let's add fs_heap_asprintf?

Comment (crafcat7) [reply fs/inode/fs_inodesearch.c:351]: Done

File: fs/driver/fs_blockproxy.c (hunk 1 above)
Comment (crafcat7) [reply fs/driver/fs_blockproxy.c:115]: Done

File: fs/inode/inode.h [hunk 3]
```diff
@@ -39,6 +39,8 @@
 #include <nuttx/fs/fs.h>
 #include <nuttx/lib/lib.h>
 
+#include "../fs_heap.h"
```
Comment (xiaoxiang781216) [inline fs/inode/inode.h:42]: remove ..

**Review by xiaoxiang781216 — APPROVED** [review]

File: fs/inode/inode.h (hunk 3 above)
Comment (crafcat7) [reply fs/inode/inode.h:42]: When referencing inode from a file in a non-fs directory, only the reference path to fs/inode is added. Therefore, if #include "fs_inode.h" is used, the file cannot be searched. This is because the fs directory is not included in the search path. like rpmsgblk_server.c

```
Note: bear 2.4.3 in Ubuntu 20.04 works out of box.
No configuration change.
make: Entering directory '/home/crafcat7/SSD/vela/nuttx.git/nuttx'
CP:  /home/crafcat7/SSD/vela/nuttx.git/nuttx/include/nuttx/config.h
CP:  /home/crafcat7/SSD/vela/nuttx.git/nuttx/include/nuttx/fs/hostfs.h
CC:  libmad/stream.c In file included from misc/rpmsgblk_server.c:34:
/home/crafcat7/SSD/vela/nuttx.git/nuttx/fs/inode/inode.h:42:10: fatal error: fs_heap.h: No such file or directory
   42 | #include <fs_heap.h>
      |          ^~~~~~~~~~~
```

File: fs/fs_heap.h [hunk 4]
```diff
@@ -25,6 +25,8 @@
  * Included Files
  ****************************************************************************/
 
+#include <stdio.h>
+#include <string.h>
```
Comment (pkarashchenko) [inline fs/fs_heap.h:29]: Does not really brings impact, but would be good to include this after
```
#include <nuttx/config.h>
#include <nuttx/compiler.h>
```

File: include/string.h [hunk 5]
```diff
@@ -110,6 +110,14 @@ FAR void  *memmem(FAR const void *haystack, size_t haystacklen,
 void explicit_bzero(FAR void *s, size_t n);
 int timingsafe_bcmp(FAR const void *b1, FAR const void *b2, size_t n);
 
+#ifdef __KERNEL__
+#define strdup(s)       nx_strdup(s)
+#define strndup(s,sz)   nx_strndup(s,sz)
```
Comment (pkarashchenko) [inline include/string.h:115]:
```suggestion
#  define strdup(s)     nx_strdup(s)
#  define strndup(s,sz) nx_strndup(s,sz)
```

File: libs/libc/stdio/lib_vasprintf.c [hunk 6]
```diff
@@ -34,6 +34,101 @@
...
+
+  lib_nulloutstream(&nulloutstream);
+  lib_vsprintf(&nulloutstream, fmt, ap);
+
+  /* Then allocate a buffer to hold that number of characters, adding one
+   * for the null terminator.
+   */
+
+  buf = kmm_malloc(nulloutstream.nput + 1);
+  if (!buf)
```
Comment (pkarashchenko) [inline libs/libc/stdio/lib_vasprintf.c:91]: Optional
```suggestion
  if (buf == NULL)
```

File: libs/libc/string/lib_strdup.c [hunk 7]
```diff
@@ -47,3 +47,16 @@ FAR char *strdup(FAR const char *s)
 
   return news;
 }
+
+FAR char *nx_strdup(FAR const char *s)
+{
+  size_t size = strlen(s) + 1;
+  FAR char *news = (FAR char *)kmm_malloc(size);
+
+  if (news)
```
Comment (pkarashchenko) [inline libs/libc/string/lib_strdup.c:56]: Optional
```suggestion
  if (news != NULL)
```

File: libs/libc/string/lib_strndup.c [hunk 8]
```diff
@@ -34,6 +34,47 @@
...
+ *   'size' is larger than the length of 's', all bytes in s will be
+ *   copied into the new memory buffer, including the terminating NUL
+ *   character. The newly created string will always be properly
+ *   terminated.
+ *
+ ****************************************************************************/
+
+FAR char *nx_strndup(FAR const char *s, size_t size)
+{
+  FAR char *news = NULL;
```
Comment (pkarashchenko) [inline libs/libc/string/lib_strndup.c:56]: Minor: `NULL` init is redundant since is overwritten unconditionally few lines below

File: libs/libc/string/lib_strndup.c [hunk 9]
```diff
@@ -34,6 +34,47 @@
...
+
+  /* Get the size of the new string (limited to size) */
+
+  size_t allocsize = strnlen(s, size);
+
+  /* Allocate the new string, adding 1 for the NUL terminator */
+
+  news = (FAR char *)kmm_malloc(allocsize + 1);
+  if (news)
```
Comment (pkarashchenko) [inline libs/libc/string/lib_strndup.c:65]: Optional
```suggestion
  if (news != NULL)
```

File: libs/libc/stdio/lib_vasprintf.c [hunk 10]
```diff
@@ -34,6 +34,101 @@
...
+   * fail. This is a known issue with x86_64.
+   */
+
+#ifdef va_copy
+  va_list ap2;
+#endif
+  FAR char *buf;
+  int nbytes;
+
+  DEBUGASSERT(ptr && fmt);
```
Comment (pkarashchenko) [inline libs/libc/stdio/lib_vasprintf.c:73]: Optional
```suggestion
  DEBUGASSERT(ptr != NULL && fmt != NULL);
```

File: include/string.h (hunk 5 above)
Comment (crafcat7) [reply include/string.h]: Done

File: libs/libc/string/lib_strdup.c (hunk 7 above)
Comment (crafcat7) [reply libs/libc/string/lib_strdup.c]: Done

File: libs/libc/string/lib_strndup.c (hunk 8 above)
Comment (crafcat7) [reply libs/libc/string/lib_strndup.c]: I removed the `NULL` initialization.

File: libs/libc/string/lib_strndup.c (hunk 9 above)
Comment (crafcat7) [reply libs/libc/string/lib_strndup.c]: Done

File: libs/libc/stdio/lib_vasprintf.c (hunk 10 above)
Comment (crafcat7) [reply libs/libc/stdio/lib_vasprintf.c]: Done

**Review by xiaoxiang781216 — APPROVED** [review]
